src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,885 @@
+                                if (!(permittedJumps & (SyntaxKind.BreakStatement ? PermittedJumps.Break : PermittedJumps.Continue))) {
+                                    // attempt to break or continue in a forbidden context
+                                    (errors || (errors = [])).push(createDiagnosticForNode(n, Messages.CannotExtractRangeContainingConditionalBreakOrContinueStatements));
+                                }
+                            }
+                            break;
+                        }
+                    case SyntaxKind.AwaitExpression:
+                        facts |= RangeFacts.IsAsyncFunction;
+                        break;
+                    case SyntaxKind.YieldExpression:
+                        facts |= RangeFacts.IsGenerator;
+                        break;
+                    case SyntaxKind.ReturnStatement:
+                        if (permittedJumps & PermittedJumps.Return) {
+                            facts |= RangeFacts.HasReturn;
+                        }
+                        else {
+                            (errors || (errors = [])).push(createDiagnosticForNode(n, Messages.CannotExtractRangeContainingConditionalReturnStatement));
+                        }
+                        break;
+                    default:
+                        forEachChild(n, visit);
+                        break;
+                }
+
+                permittedJumps = savedPermittedJumps;
+            }
+        }
+    }
+
+    /**
+     * Computes possible places we could extract the function into. For example,
+     * you may be able to extract into a class method *or* local closure *or* namespace function,
+     * depending on what's in the extracted body.
+     */
+    export function collectEnclosingScopes(range: TargetRange): Scope[] {
+        let current: Node = isArray(range.range) ? firstOrUndefined(range.range) : range.range;
+        if (range.facts & RangeFacts.UsesThis) {
+            // if range uses this as keyword or as type inside the class then it can only be extracted to a method of the containing class
+            const containingClass = getContainingClass(current);
+            if (containingClass) {
+                return [containingClass];
+            }
+        }
+
+        const scopes: Scope[] = [];
+        while (current) {
+            // We want to find the nearest parent where we can place an "equivalent" sibling to the node we're extracting out of.
+            // Walk up to the closest parent of a place where we can logically put a sibling:
+            //  * Function declaration
+            //  * Class declaration or expression
+            //  * Module or source file
+            // Note that we don't use isFunctionLike because we don't want to put the extracted closure *inside* a method
+            if ((current.kind === SyntaxKind.FunctionDeclaration) || isSourceFile(current) || isModuleBlock(current) || isClassLike(current)) {

```

Comment:
The body of a namespace is `SyntaxKind.ModuleBlock` (the `namespace` keyword is an alias for the `module` keyword, hence the naming confusion)